librarian/data/meta/archive.py

```diff
@@ -0,0 +1,673 @@
+                self._tasks.schedule(self.scan, kwargs=kwargs, delay=delay)
+            else:
+                for metadata in self._scan(**kwargs):
+                    yield metadata
+
+    def scan(self, path=None, partial=False, callback=None, maxdepth=None,
+             depth=0, delay=0):
+        """
+        Traverse ``path`` and py:meth:`~Archive.analyze` all encountered files.
+
+        In case ``callback`` is specified, the traversing will be performed
+        asynchronously, with each next level scheduled as a separate background
+        task, and invoking ``callback`` with each result set separately. If
+        ``callback`` was not specified, it will behave as an iterator, yielding
+        the results of scan on each level. ``delay`` is used only in async mode
+        and it represents the amount of seconds before the next directory is to
+        be scanned. ``maxdepth`` can limit how deep the scan is allowed to
+        traverse directory tree.
+        """
+        generator = self._scan(path, partial, callback, maxdepth, depth, delay)
+        if callback:
+            # evaluate generator(because no-one else will) so that ``callback``
+            # actually gets executed
+            return list(generator)
+        # return unevaluated generator
+        return generator
+
+    def _keep_supported(self, paths, content_type):
+        """
+        Return a list of only those ``paths`` that belong to ``content_type``.
+        """
+        processor_cls = self.Processor.for_type(content_type)
+        return [path for path in paths if processor_cls.can_process(path)]
+
+    def _strip(self, metadata, content_type=None):
+        """
+        Return a copy of the passed in ``metadata`` with those keys stripped
+        out that do not belong to the specified ``content_type``.
+        """
+        keys = self.ContentTypes.keys(content_type)
+        return dict((k, v) for (k, v) in metadata.items() if k in keys)
+
+    def _reconstruct_meta(self, rows):
+        """
+        Return py:attr:`~Archive.MetaWrapper` instances created from ``rows``.
+
+        ``rows`` is expected to be ordered by the ``path`` column.
+        """
+        path = data = None
+        for row in rows:
+            # path changed, so row belongs to a different fs entry
+            if row['path'] != path:
+                # yield previous data element before creating the new one
+                if data:
+                    yield self.MetaWrapper(data)
+                # create next fs object
+                path = row['path']
+                data = dict(path=path,
+                            id=row['id'],
+                            parent_id=row['parent_id'],
+                            type=row['type'],
+                            mime_type=row['mime_type'],
+                            content_types=row['content_types'])
+            # put meta key/value pairs into fs data element
+            key = row['key']
+            if key:
```

Comment:
Yes, some of the right joins where no meatadata was found, only the fs entries result in the meta columns having null values